cpp/src/slice2swift/SwiftUtil.cpp

```diff
@@ -531,7 +531,21 @@ SwiftGenerator::validateMetadata(const UnitPtr& u)
         .validOn = {typeid(Module)},
         // Even though it's really 'module:prefix' the validator sees this as a single argument since there's no commas.
         .acceptedArgumentKind = MetadataArgumentKind::SingleArgument,
-    };
+        .extraValidation = [](const MetadataPtr& metadata, const SyntaxTreeBasePtr& p) -> optional<string>
+        {
+            const string msg = "'swift:module' is deprecated; use 'swift:identifier' to remap modules instead";
+            p->unit()->warning(metadata->file(), metadata->line(), Deprecated, msg);
+
+            if (auto contained = dynamic_pointer_cast<Contained>(p))
+            {
+                if (contained->hasMetadata("swift:identifier"))
+                {
+                    return "the 'swift:module' metadata cannot be used alongside 'swift:identifier' - both change the "
+                           "mapped name of this module";
+                }
```

Comment:
... unless a user used 'swift:module' and 'swift:identifier' together, in which case we return an error message.
The MetadataValidation machinery will handle the rest for us.